http/src/lib.rs

```diff
@@ -235,78 +182,177 @@ impl<M: jsonrpc::Metadata, S: jsonrpc::Middleware<M>> ServerBuilder<M, S> {
+		})?;
+
 		Ok(Server {
-			server: Some(l),
+			addr: [local_addr],
 			handle: Some(handle),
-			event_loop: event_loop,
+			close: Some(close),
 		})
 	}
 }
 
+struct PanicHandler {
+	handler: Option<Box<Fn() -> () + Send>>,
+}
+
+impl Drop for PanicHandler {
+	fn drop(&mut self) {
+		if ::std::thread::panicking() {
+			if let Some(ref h) = self.handler {
+				h();
+			}
+		}
+	}
+}
+
+/// Tokio-proto JSON-RPC HTTP Service
+pub struct RpcService<M: jsonrpc::Metadata, S: jsonrpc::Middleware<M>> {
+	handler: Arc<MetaIoHandler<M, S>>,
+	meta_extractor: Arc<HttpMetaExtractor<M>>,
+	hosts: Option<Vec<String>>,
+	cors_domains: Option<Vec<cors::AccessControlAllowOrigin>>,
+}
+
+fn is_json(content_type: Option<&str>) -> bool {
+	match content_type {
+		None => false,
+		Some(ref content_type) => {
+			let json = "application/json";
+			content_type.eq_ignore_ascii_case(json)
+		}
+	}
+}
+
+impl<M: jsonrpc::Metadata, S: jsonrpc::Middleware<M>> tokio_service::Service for RpcService<M, S> {
+	type Request = tokio_minihttp::Request;
+	type Response = tokio_minihttp::Response;
+	type Error = io::Error;
+	// TODO avoid boxing here
+	type Future = BoxFuture<tokio_minihttp::Response, io::Error>;
+
+	fn call(&self, request: Self::Request) -> Self::Future {
+		let request = req::Req::new(request);
+		// Validate HTTP Method
+		let is_options = request.method() == req::Method::Options;
+		if !is_options && request.method() != req::Method::Post {
+			return future::ok(
+				res::method_not_allowed()
+			).boxed();
+		}
+
+		// Validate allowed hosts
+		let host = request.header("Host");
+		if !hosts::is_host_valid(host, &self.hosts) {
+			return future::ok(
+				res::invalid_host()
+			).boxed();
+		}
+
+		// Validate content type
+		let content_type = request.header("Content-type");
+		if !is_json(content_type) {
+			return future::ok(
+				res::invalid_content_type()
+			).boxed();
+		}
+
+		// Extract CORS headers
+		let origin = request.header("Origin");
+		let cors = cors::get_cors_header(origin, &self.cors_domains);
+
+		// Don't process data if it's OPTIONS
+		if is_options {
+			return future::ok(
+				res::new("", cors)
+			).boxed();
+		}
+
+		// Extract metadata
+		let metadata = self.meta_extractor.read_metadata(&request);
+
+		// Read & handle request
+		let data = request.body();
+		self.handler.handle_request(data, metadata).map(|result| {
+			let result = format!("{}\n", result.unwrap_or_default());
+			res::new(&result, cors)
+		}).map_err(|_| unimplemented!()).boxed()
+	}
+}
+
 /// jsonrpc http server instance
 pub struct Server {
-	server: Option<server::Listening>,
+	addr: [SocketAddr; 1],
 	handle: Option<thread::JoinHandle<()>>,
-	event_loop: Option<RpcEventLoopHandle>,
+	close: Option<futures::sync::oneshot::Sender<()>>,
 }
 
 impl Server {
 	/// Returns addresses of this server
 	pub fn addrs(&self) -> &[SocketAddr] {
-		self.server.as_ref().unwrap().addrs()
+		&self.addr
```

Comment:
It's no longer "addresses". Couldn't we just replace the `[SocketAddr; 1]` with a `SocketAddr` and return that? Is there a lot of API breakage?